### PR TITLE
Initial access logs implementation

### PR DESCRIPTION
**Pull request by howardjohn**
This PR adds "Access logs". There are logs emitted in three cases:
1. Premature failure (invalid src/dst, etc)
2. Connection start (debug level)
3. Connection end

These emit a variety of info about the result, as well as src/destination information.

This PR introduces a *ton* of changes outside of logging. The primary reason around this is refactoring to allow us to accurately and safely emit a log at the appropriate times (notably, at the end of the request). Things like `?` short-circuit our ability to record.  So there is a lot of refactoring around that. Additionally, some RBAC logic is deferred until we have proper src/dst metadata, so we can record that in telemetry if RBAC is denied; I don't see any reason this would be an issue.

Example logs (not some are `debug` -- these are not visble by default)

```
2024-04-11T00:26:42.295630Z DEBUG access: connection opened src.addr=shell-6d8bcd654d-t88gp:59690 src.workload="shell" src.namespace="default" src.identity="spiffe://cluster.local/ns/default/sa/default" dst.addr=namespace-istio-waypoint-9ccc6c96-hthtk:15008 dst.hbone_addr="echo:80" direction="outbound"
2024-04-11T00:26:42.309245Z DEBUG access: connection opened src.addr=namespace-istio-waypoint-9ccc6c96-hthtk:36782 src.workload="namespace-istio-waypoint" src.namespace="default" src.identity="spiffe://cluster.local/ns/default/sa/namespace-istio-waypoint" dst.addr=echo-8659868fcc-z2mvm:80 dst.hbone_addr="echo-8659868fcc-z2mvm:80" dst.workload="echo" dst.namespace="echo" dst.identity="spiffe://cluster.local/ns/default/sa/default" direction="inbound"
2024-04-11T00:26:42.311513Z  INFO access: connection complete src.addr=shell-6d8bcd654d-t88gp:59690 src.workload="shell" src.namespace="default" src.identity="spiffe://cluster.local/ns/default/sa/default" dst.addr=namespace-istio-waypoint-9ccc6c96-hthtk:15008 dst.hbone_addr="echo:80" direction="outbound" bytes_sent=67 bytes_recv=490 duration="16ms"
2024-04-11T00:26:47.314293Z  INFO access: connection complete src.addr=namespace-istio-waypoint-9ccc6c96-hthtk:36782 src.workload="namespace-istio-waypoint" src.namespace="default" src.identity="spiffe://cluster.local/ns/default/sa/namespace-istio-waypoint" dst.addr=echo-8659868fcc-z2mvm:80 dst.hbone_addr="echo-8659868fcc-z2mvm:80" dst.workload="echo" dst.namespace="echo" dst.identity="spiffe://cluster.local/ns/default/sa/default" direction="inbound" bytes_sent=399 bytes_recv=144 duration="5005ms"
```

This depends on https://github.com/istio/ztunnel/pull/915 because doing it without that is a major annoyance. But if we need to keep it, I can update this PR to allow it.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -100,6 +100,7 @@ tracing-subscriber = { version = "0.3", features = ["registry", "env-filter", "j
 url = "2.2"
 x509-parser = { version = "0.16", default-features = false }
 backoff = "0.4.0"
+tracing-core = "0.1.32"
```
Comment (costinm) [inline Cargo.toml:103]: Is this compatible with OTLP ?

Comment (howardjohn) [reply Cargo.toml:103]: Its just the `tracing` library, so its pluggable. Right now we don't have code to do it, but we could.

Even better, when we do we could easily choose to send JUST access logs (or send them to a different place, handle them differently, whatever)

File: src/proxy/metrics.rs [hunk 2]
```diff
@@ -362,15 +363,36 @@ impl Recorder<BytesTransferred<'_>, (u64, u64)> for super::Metrics {
...
+            self.sent_bytes.get_or_create(&tl).inc_by(sent);
         }
         if recv != 0 {
-            self.received_bytes
-                .get_or_create(&CommonTrafficLabels::from(event.0))
-                .inc_by(recv);
+            self.received_bytes.get_or_create(&tl).inc_by(recv);
         }
+        let mtls = tl.connection_security_policy == SecurityPolicy::mutual_tls;
+        // TODO: result status
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply src/proxy/metrics.rs:374]: Agree, working on this now. Needs a lot of refactoring though

File: src/proxy/inbound.rs [hunk 3]
```diff
@@ -173,131 +181,79 @@ impl Inbound {
...
-                            .increment_defer::<_, metrics::ConnectionClose>(&connection_metrics);
-
-                        let _extra_conn_close = extra_connection_metrics
-                            .as_ref()
-                            .map(|co| metrics.increment_defer::<_, metrics::ConnectionClose>(co));
-
-                        let transferred_bytes =
-                            metrics::BytesTransferred::from(&connection_metrics);
-                        match request_type {
-                            DirectPath(mut incoming) => {
```
Comment (bleggett) [inline src/proxy/inbound.rs:214]: Ditching `DirectXXX`? I don't mind since they're unused ATM, I guess.

**Review by bleggett — APPROVED** [review]
Good cleanups, ty

**Review by ilrudie — COMMENTED** [review]
Over all this seems good. Couple of minor comments but I didn't notice any show stoppers

File: src/proxy/metrics.rs [hunk 4]
```diff
@@ -346,31 +330,183 @@ impl Recorder<ConnectionOpen, u64> for Metrics {
...
+            );
+        } else {
+            event!(
+                target: "access",
+                parent: None,
+                tracing::Level::WARN,
+
+                src.addr = %self.src,
+                src.workload = tl.source_canonical_service.inner(),
+                src.namespace = tl.source_workload_namespace.inner(),
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply src/proxy/metrics.rs]: The level is a const. So I think we could but would need a macro unfortunately

[the remark this answers is not in the document]
Comment (ilrudie) [reply src/proxy/metrics.rs]: macro in macro seems a bit much😬

[the remark this answers is not in the document]
Comment (howardjohn) [reply src/proxy/metrics.rs]: I switched to the macro. Its not _that_ bad, and I already diverged on the paths ~5 times in development accidentally, so probably nice to have

[the remark this answers is not in the document]
Comment (ilrudie) [reply src/proxy/metrics.rs]: ah, this works well because you did it in a decidedly smarter way than I was thinking initially 😀

**Review by stevenctl — APPROVED** [review]

**Pull request merged**